Do not merge | Looking for advice on listener startup robustness improvements

ambrosejcarr opened the pull request:
**This is not urgent, and can be tabled until after the demo if desired.**

This PR makes some changes to how checks are run when the listener is launched. Specifically, 
it checks the config.json file for:
- [x] duplicate wdl specifications
- [x] fields missing in wdl specifications
- [x] unexpected (extra) fields in wdl specifications
- [x] fields pointing to non-existent google storage endpoints

Everything works when I limit the testing to the first three types. However, there are some problems with the last one. The most significant is that at the time of setup inside the docker environment, I don't believe `gcloud auth` has been run (I had to add `google-cloud` to `requirements.txt`), so the listener fails to start after throwing a google-auth error (see error log below).

**My main question is: am I missing something and using the wrong config method? If not, do we *want* to config this early? I see benefit in these tests, but it might not be worth it to run this early config.**

```
<ajc@:/Users/ajc/projects/secondary-analysis> ./run_docker.sh dev test
Updated property [core/project].
Activated service account credentials for: [devtest-gcs-reader@broad-dsde-mint-dev.iam.gserviceaccount.com]
DEBUG:connexion.app:Root Path: /secondary-analysis/green_box
DEBUG:connexion.app:Specification directory: /secondary-analysis/green_box
DEBUG:connexion.app:Setting error handlers
DEBUG:google.auth.transport._http_client:Making request: GET http://169.254.169.254
INFO:google.auth.compute_engine._metadata:Compute Engine Metadata server unavailable.
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/secondary-analysis/green_box/__init__.py", line 121, in <module>
    app = main()
  File "/secondary-analysis/green_box/__init__.py", line 112, in main
    verify_config_json(config)
  File "/secondary-analysis/green_box/__init__.py", line 96, in verify_config_json
    client = storage.Client()  # this is slow, do this only once
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/storage/client.py", line 59, in __init__
    _http=_http)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/client.py", line 211, in __init__
    _ClientProjectMixin.__init__(self, project=project)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/client.py", line 165, in __init__
    project = self._determine_default(project)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/client.py", line 178, in _determine_default
    return _determine_default_project(project)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/_helpers.py", line 179, in _determine_default_project
    _, project = google.auth.default()
  File "/usr/local/lib/python2.7/dist-packages/google/auth/_default.py", line 286, in default
    raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or
explicitly create credential and re-run the application. For more
information, please see
https://developers.google.com/accounts/docs/application-default-credentials.
```

In order to make unit testing of functions in `green_box.__init__` possible, I restructured the `green_box` folder slightly:  I included an additional `__init__.py` files to cause python to view `green_box` as a package. This does not influence the above error, but it has the corollary of removing the `green-box-api.py` file. 

The unit tests themselves all run successfully. From within `secondary-analysis` you can run:
`python2.7 -m green_box.test.test_startup`

the startup procedure now induces `start_listener.sh` to directly invoke `green_box.__init__`. These changes go a ways towards changing `green_box` into an installable package; it's just missing a `setup.py` file and a small refactor to move `green_box.__init__.main()` into `green_box.main`

I'm also open to any/all criticism of the actual code. Please note that a few of the changes in `notifications.py` were just adjusting the file to be `pep8` compliant (formatting changes).

dshiga commented on green_box/__init__.py:
```diff
@@ -1,26 +1,122 @@
+            'have identical configuration. Please check configuration file contents.')
+
+    # check that link fields point to valid gs endpoints
+    client = storage.Client()  # this is slow, do this only once
```
I think you need to create the client using from_service_account_json, as described here:
https://google-cloud-python.readthedocs.io/en/latest/core/auth.html#service-accounts

There is a key file available here: /etc/secondary-analysis/bucket-reader-key.json
(listener-start.sh is using it)

dshiga commented on green_box/__init__.py:
```diff
@@ -1,26 +1,122 @@
-#!/usr/bin/env python
+#!/usr/bin/env/python
```
I don't think we want to remove that space. With the space, it's calling env and passing "python" as an argument to it. Without it, it's going to try to find and run python in a directory called /usr/bin/env.

ambrosejcarr replied:
Didn't realize this! thanks for the explanation.

dshiga commented on green_box/api/notifications.py:
```diff
@@ -59,31 +66,34 @@ def post(body):
     response.headers['Content-type'] = 'application/json'
     return response
 
+
```
Interesting, does PEP8 suggest two empty lines before each function?

ambrosejcarr replied:
Yea; two spaces between *top-level* functions and classes.

dshiga commented on green_box/__init__.py:
```diff
@@ -1,26 +1,122 @@
 
-resolver = RestyResolver("green_box.api", collection_endpoint_name="list")
-app.add_api('../green_box.yml', resolver=resolver, validate_responses=True)
+if __name__ == "__main__":
```
I like this! Making this callable directly is nice.

dshiga commented on green_box/__init__.py:
```diff
@@ -1,26 +1,122 @@
+            '{blob_name}'.format(bucket=bucket, blob_name=blob_name))
+
+
+def verify_config_json(parsed_json_dict):
```
maybe a different name for this? parsed_json_dict made me think you'd converted it into something other than json

ambrosejcarr replied:
Changed, thanks.

dshiga commented on green_box/__init__.py:
```diff
@@ -1,26 +1,122 @@
+    """
+
+    # check for missing or unexpected config fields
+    expected_fields = {
```
I wonder if it might be helpful to define something like a WdlConfig or SubscriptionConfig class for the checks you want to do here.

ambrosejcarr replied:
Added but not yet integrated; want to fix permissions error first.

dshiga commented on green_box/__init__.py:
```diff
@@ -1,26 +1,122 @@
+    if len(parsed_json_dict['wdls']) != len(set(subscription_ids)):
+        raise ValueError(
+            'One or more wdl specifications contains a duplicated subscription ID but '
+            'have identical configuration. Please check configuration file contents.')
```
I think you already checked for completely identical configs above, so here you've actually found identical subscription ids but with different contents, right?

ambrosejcarr replied:
Yep, forgot to add "not". Good catch.

ambrosejcarr commented:
Thank you very much for the suggestions, Dave. I've implemented each of your suggestions including the listener authentication. 

Unfortunately, I am still getting an error. Perhaps I have an early version of the key that doesn't have proper permissions? 

I've confirmed that my docker build has the key in the correct location and that the contents match the key on my system. I get the same error when I run the test locally from the same key that gets copied to the docker instance. 

```
<ajc@:/Users/ajc/projects/secondary-analysis> ./run_docker.sh dev test
Updated property [core/project].
Activated service account credentials for: [devtest-gcs-reader@broad-dsde-mint-dev.iam.gserviceaccount.com]
Configuring listener credentials using /etc/secondary-analysis/bucket-reader-key.json
Configuring listener credentials using /etc/secondary-analysis/bucket-reader-key.json
DEBUG:connexion.app:Root Path: /secondary-analysis/green_box
DEBUG:connexion.app:Specification directory: /secondary-analysis/green_box
DEBUG:connexion.app:Setting error handlers
DEBUG:google.auth.transport.requests:Making request: POST https://accounts.google.com/o/oauth2/token
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): accounts.google.com
DEBUG:urllib3.connectionpool:https://accounts.google.com:443 "POST /o/oauth2/token HTTP/1.1" 200 None
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): www.googleapis.com
DEBUG:urllib3.connectionpool:https://www.googleapis.com:443 "GET /storage/v1/b/broad-dsde-mint-dev-teststorage?projection=noAcl HTTP/1.1" 403 440
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/secondary-analysis/green_box/__init__.py", line 290, in <module>
    app = main()
  File "/secondary-analysis/green_box/__init__.py", line 281, in main
    verify_config_json(config)
  File "/secondary-analysis/green_box/__init__.py", line 269, in verify_config_json
    verify_gs_link(value)
  File "/secondary-analysis/green_box/__init__.py", line 54, in verify_gs_link
    verified_bucket = client.get_bucket(bucket)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/storage/client.py", line 173, in get_bucket
    bucket.reload(client=self)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/storage/_helpers.py", line 99, in reload
    _target_object=self)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/_http.py", line 293, in api_request
    raise exceptions.from_http_response(response)
google.cloud.exceptions.Forbidden: 403 GET https://www.googleapis.com/storage/v1/b/broad-dsde-mint-dev-teststorage?projection=noAcl: devtest-gcs-reader@broad-dsde-mint-dev.iam.gserviceaccount.com does not have storage.buckets.get access to bucket broad-dsde-mint-dev-teststorage.
```

dshiga commented on green_box/SubscriptionConfig.py:
```diff
@@ -0,0 +1,158 @@
+        ]))
+
+
+class SubscriptionConfig:
```
I would probably call this one ListenerConfig. It's really the top level configuration for the app, which we're calling the listener at the moment.

ambrosejcarr commented:
The culprit: 

```
In [41]: bucket = client.bucket('broad-dsde-mint-dev-teststorage')

In [42]: next(iter(bucket.list_blobs()))
Out[42]: <Blob: broad-dsde-mint-dev-teststorage, 10x/benchmark/1e4/pbmc8k_S1_L007_I1_001.fastq.gz>

In [43]: blob = next(iter(bucket.list_blobs()))

In [44]: blob.exists()  # works, hurray; this is what I need
Out[44]: True

In [45]: bucket.exists()  # this is a check I do earlier in the workflow.. because it makes sense to do..
---------------------------------------------------------------------------
Forbidden                                 Traceback (most recent call last)
<ipython-input-45-3fc085c273f9> in <module>()
----> 1 bucket.exists()

/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/google/cloud/storage/bucket.pyc in exists(self, client)
    170             client._connection.api_request(
    171                 method='GET', path=self.path,
--> 172                 query_params=query_params, _target_object=None)
    173             # NOTE: This will not fail immediately in a batch. However, when
    174             #       Batch.finish() is called, the resulting `NotFound` will be

/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/google/cloud/_http.pyc in api_request(self, method, path, query_params, data, content_type, headers, api_base_url, api_version, expect_json, _target_object)
    291
    292         if not 200 <= response.status_code < 300:
--> 293             raise exceptions.from_http_response(response)
    294
    295         if expect_json and response.content:

Forbidden: 403 GET https://www.googleapis.com/storage/v1/b/broad-dsde-mint-dev-teststorage?fields=name: devtest-gcs-reader@broad-dsde-mint-dev.iam.gserviceaccount.com does not have storage.buckets.get access to bucket broad-dsde-mint-dev-teststorage.
```

I should be able to fix this with a work-around. 

**The take-home is that read_only scoping (which was required) does not let you list, check, or manipulate buckets**

samanehsan commented on green_box/api/notifications.py:
```diff
@@ -43,14 +49,15 @@ def post(body):
-    wdl_deps_file = wdl['wdl_deps_file']
-    wdl_default_inputs_file = wdl['default_inputs_file']
-    cromwell_response = start_workflow(wdl_file, wdl_deps_file, 'cromwell_inputs.json', wdl_default_inputs_file, green_config)
+    wdl_file = os.path.split(wdl['wdl_link'])[-1]
```
Good call on changing this to get the file names from the links! :+1:

samanehsan commented on green_box/__init__.py:
```diff
@@ -4,12 +4,117 @@
+    logging.debug('Configuring listener credentials using %s' % key_location)
+    TEST = False
+except IOError:
+    # try:
```
Did you mean to remove this commented out code?

samanehsan commented on green_box/__init__.py:
```diff
@@ -4,12 +4,117 @@
+        key_location, scopes=scopes)
+    client = storage.Client(credentials=credentials, project=credentials.project_id)
+    logging.debug('Configuring listener credentials using %s' % key_location)
+    TEST = False
```
I think a more descriptive name would be useful here to indicate what this flag does, such as `TEST_GS_LINKS`

samanehsan commented on green_box/__init__.py:
```diff
@@ -4,12 +4,117 @@
+    """
+    # split the link, checking to make sure it contains both a bucket and key
+    # python3 syntax
+    # try:
```
Did you mean to remove this code as well?

samanehsan commented on green_box/test/test_green_box.py:
```diff
@@ -0,0 +1,107 @@
+
+# load a json string representing a real json configuration.
+correct_test_config = json.loads("""
+{
```
Could you import `test/config.json` here instead?

samanehsan commented on green_box/test/test_green_box.py:
```diff
@@ -0,0 +1,107 @@
+
+
+# todo refactor into package for easier testing
+# run tests from secondary-analysis folder with command:
```
These comments on how to run the tests would make more sense in a docstring at the top of the file or maybe there could be a section in the readme about running tests

samanehsan commented on green_box/SubscriptionConfig.py:
```diff
@@ -0,0 +1,158 @@
+        self._default_inputs_link = value
+
+    def __eq__(self, other):
+        if not isinstance(other, WdlConfig):
```
This can be simplified to `return isinstance(other, WdlConfig) and hash(self) == hash(other)`

samanehsan commented on green_box/SubscriptionConfig.py:
```diff
@@ -0,0 +1,158 @@
+        return self._wdls
+
+    @wdls.setter
+    def wdls(self, wdl_configs):
```
This checks the `wdl_configs`, but still need to set them

samanehsan commented:
Just left a few comments on your PR @ambrosejcarr! 🌤

ambrosejcarr commented:
@samanehsan Thanks for the comments; Dave requested that I leave the google-storage checks behind due to the code burden they add; I'm redoing this with the classes and have incorporated your comments into that branch.
